packages/compiler/lib/src/compilation/hir/expressions.dart

```diff
@@ -89,6 +94,7 @@ abstract class Expression implements _$Expression {
         while_: (_, __, ___, type) => type,
         break_: (_, __, ___) => CandyType.never,
         continue_: (_, __) => CandyType.never,
+        assignment: (_, left, __) => left.type,
```

Comment:
An assignment expression should evaluate to its value.
```suggestion
        assignment: (_, left, __) => right.type,
```